[ASM] - Expander - Update Service Ownership

kball-pa opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Description
- Updated Cortex ASM - GCP Enrichment playbook to retrieve the service account on the instance and writes it to incident field **asmserviceownerunrankedraw** for use by the **Cortex ASM - Service Ownership** playbook.
- Updated the Service Ownership playbook to run for all cloud service providers and added support for retrieving GCP project owners from user-managed, cross-project service accounts.
- Add script GetProjectOwners which prses a GCP service account email for the project ID, then looks up the project owners and adds them to a list of potential service owners for ranking.
- Updated script RankServiceOwners to no longer limit to the top-5 service owners

## Minimum version of Cortex XSOAR
- 6.8.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x ] No

## Must have
- [x] Tests
- [x] Documentation

BigEasyJ commented:
@thefrieddan1 @efelmandar I'm seeing `The core pack CortexAttackSurfaceManagement cannot depend on non-core packs: ['GCP-IAM'] - revert this change.` in the validation error and usually that is indicative of `skipavailable: false` on a task, however I do not see a tasks that would cause this.

kball-pa commented:
@thefrieddan1  @efelmandar Adding to John's comment above, I also tried setting `skipunavailable: true` for all non-conditional tasks that I added to each of the playbooks I modified, and still was getting the same error with local validation (`The core pack CortexAttackSurfaceManagement cannot depend on non-core packs: ['GCP-IAM'] - revert this change.`)

johnnywilkes commented:
@kball-pa , I looked into that validation error and not sure why you are getting it.  Can you remove all your changes outside the ASM pack for this PR so that we can approve, please?

BigEasyJ commented:
@johnnywilkes Kelsey's branch does not changes outside the ASM pack for this PR
![Screenshot 2023-06-02 at 1 16 29 PM](https://github.com/demisto/content/assets/40349459/d40405f8-58fd-4ec6-a925-7a9e93fe22c5)

kball-pa commented:
Even when I set the base branch to be my contributor branch on a separate diff, it only shows changes to the ASM pack. Not sure why it's different on this PR.

![Screenshot 2023-06-02 at 1 19 03 PM](https://github.com/demisto/content/assets/131012047/cd407c2e-cd6e-4c25-9029-9d0c44f60246)

johnnywilkes commented:
@kball-pa / @BigEasyJ , apologies if it was an issue on my end, looks good now

BigEasyJ commented:
This is a GH bug.

Fix: Use the Edit button when viewing the PR in GitHub to the change into master. Then switch it back to contrib branch and it will now correctly show only the changes from the most recent commits.

BigEasyJ reviewed (changes requested):
Remember to update the pack ReadMe with your new script.

johnnywilkes commented on Packs/CortexAttackSurfaceManagement/Scripts/GetProjectOwners/GetProjectOwners.py:
```diff
@@ -0,0 +1,100 @@
+    try:
+        return demisto.executeCommand("gcp-iam-project-iam-policy-get", args={"project_name": f"projects/{project_name}"})
+    except Exception as ex:
+        raise RuntimeError(f"Error retrieving IAM policy for GCP project {project_name}") from ex
```
we will have to treat this differently because it could cause playbooks to fail where a user doesn't have GCP-IAM integration setup.  We should check for integration i playbook or add some handling in here and maybe just return to war room some like like "Error retrieving IAM policy for GCP project {project_name}" instead of Exception.

ptoman-pa commented on Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py:
```diff
@@ -119,11 +117,14 @@ def aggregate(owners: List[Dict[str, str]]) -> List[Dict[str, Any]]:
         for other in other_keys:
             if keys_to_types[other] == str:
                 # union over strings
-                owner[other] = ' | ' .join(sorted(
+                owner[other] = ',' .join(sorted(
```
Can we retain " | " across the board (instead of ",")?

I ask because "," is much more frequent, so it's much more fragile to parse independent sources from each other  downstream. Specifically, the following becomes much more dangerous/no longer reasonable: `owner.get("sources", "").split(",")`.

kball-pa replied:
That's fine with me. I think our plan is to make `Justification` be a more human-readable string, so maybe the UI can read from that field instead of `Source`. Alternatively if downstream users want a comma-separated list of sources they can `.replace(' | ', ',')`

@BigEasyJ what do you think?

kball-pa commented on Packs/CortexAttackSurfaceManagement/Playbooks/Cortex_ASM_-_GCP_Enrichment_README.md:
```diff
@@ -38,4 +38,4 @@ There are no outputs for this playbook.
 
-![Cortex ASM - GCP Enrichment](../doc_files/Cortex_ASM_-_GCP_Enrichment.png)
\ No newline at end of file
+![Cortex ASM - GCP Enrichment](../doc_files/Cortex_ASM_-_GCP_Enrichment.png)
```
i'm guessing there was trailing whitespace committed previously -- my editor is configured to auto-trim whitespace

BigEasyJ commented on Packs/CortexAttackSurfaceManagement/Scripts/GetProjectOwners/GetProjectOwners.py:
```diff
@@ -0,0 +1,100 @@
+    Retrieve IAM policy for project
+    """
+    try:
+        return demisto.executeCommand("gcp-iam-project-iam-policy-get", args={"project_name": f"projects/{project_name}"})
```
Check for if the integration with `gcp-iam-project-iam-policy-get` is available either in the script or in the playbook. If you do it in a script it requires `demisto.getModules()` and getting the `state` and `brand`.

kball-pa replied:
tested that the error handling in main works! if there are GCP owners in the unranked list but the integration is disabled, it logs "No additional owners found" and continues the playbook

BigEasyJ reviewed (changes requested):
Looks good, just one more change.

johnnywilkes reviewed (approved):
@kball-pa Great work on this!  @thefrieddan1 this is an ASM only PR, please merge when possible
